src/odemis/driver/smaract.py

```diff
@@ -291,119 +295,203 @@ def __str__(self):
         self._metadata[model.MD_SW_VERSION] = self._swVersion
         logging.debug("Using SmarPod library version %s", self._swVersion)
 
         self.position = model.VigilantAttribute({}, readonly=True)
+        self._updatePosition()
+
+        # TODO: Change pivot mode to fixed automatically?
+        self.Set_ui(SmarPodDLL.PIVOT_MODE, SmarPodDLL.PIVOT_FIXED)
+        self._metadata[model.MD_PIVOT_POS] = self.GetPivot()
 
         # will take care of executing axis move asynchronously
         self._executor = CancellableThreadPoolExecutor(1)  # one task at a time
 
-        referenced = c_int()
-        self.core.Smarpod_IsReferenced(self._id, byref(referenced))
+        # Use a default actuator speed
+        self.SetSpeed(speed)
+        self.speed = VigilantAttribute({}, unit="m/s", readonly=True)
+        self._updateSpeed()
+        self.SetAcceleration(accel)
+        self._accel = self.GetAcceleration()
+
+        referenced = self.IsReferenced()
         # define the referenced VA from the query
-        axes_ref = {a: referenced.value for a, i in self.axes.items()}
+        axes_ref = {a: referenced for a in self.axes}
         # VA dict str(axis) -> bool
         self.referenced = model.VigilantAttribute(axes_ref, readonly=True)
         # If ref_on_init, referenced immediately.
-        if referenced.value:
+        if referenced:
             logging.debug("SmarPod is referenced")
         else:
-            logging.warning("SmarPod is not referenced. The device will not function until referencing occurs.")
             if ref_on_init:
-                self.reference().result()
-
-        # Use a default actuator speed
-        self._set_speed(actuator_speed)
-        self._speed = self._get_speed()
-        self._accel = self.GetAcceleration()
+                self.reference(set(axes_ref.keys()))  # will reference in background
+            else:
+                logging.warning("SmarPod is not referenced. The device will not function until referencing occurs.")
 
-        self._updatePosition()
+        self._update_position_timer = RepeatingTimer(1.0, self._updatePositionInBackground)
+        self._update_position_timer.start()
 
     def terminate(self):
+        self._update_position_timer.cancel()
+
+        if self._executor:
+            self.Stop()
+            self._executor.shutdown()
+            self._executor = None
+            self.core.Smarpod_Close(self._id)
+
         # should be safe to close the device multiple times if terminate is called more than once.
-        self.core.Smarpod_Close(self._id)
         super(SmarPod, self).terminate()
+
+    def updateMetadata(self, md):
+        if model.MD_PIVOT_POS in md:
+            pivot = md[model.MD_PIVOT_POS]
+            if not (isinstance(pivot, dict) and set(pivot.keys()) == {"x", "y", "z"}):
+                raise ValueError("Invalid metadata, should be a coordinate dictionary but got %s." % (pivot,))
```

Comment:
I would specify this even further: `"Invalid metadata, should be a coordinate dictionary containing 'x', 'y' and 'z' but got %s." % (pivot,)"`